spec/services/container_deployment_service_spec.rb

```diff
@@ -0,0 +1,43 @@
+require 'recursive-open-struct'
+
+describe ContainerDeploymentService do
+  before do
+    %w(amazon openstack google azure redhat vmware).each do |p|
+      network = FactoryGirl.create(:network, :ipaddress => "127.0.0.1")
+      @hardware = FactoryGirl.create(:hardware)
+      @hardware.networks << network
+      if (p.include? "amazon") || (p.include? "redhat")
+        template = FactoryGirl.create("template_#{p}".to_sym)
+        provider = FactoryGirl.create("ems_#{p}".to_sym, :miq_templates => [template])
+      else
+        provider = FactoryGirl.create("ems_#{p}".to_sym)
+      end
+      FactoryGirl.create("vm_#{p}".to_sym, :ext_management_system => provider, :hardware => @hardware)
+    end
+    @foreman_provider = FactoryGirl.create(:configuration_manager_foreman)
+  end
+
+  context "possible_providers_and_vms" do
+    it "finds all Cloud and Infra providers and their existing VMs" do
+      providers = ContainerDeploymentService.new.possible_providers_and_vms
+      vms = []
+      providers.each do |p|
+        vms << p[:vms]
+      end
+      expect(providers.size).to eq(6)
+      expect(vms.size).to eq(6)
+    end
+  end
+
+  context "possible_provision_providers" do
+    it "finds all providers with provision ability supported by deployment, and their templates" do
+      providers = ContainerDeploymentService.new.possible_provision_providers
+      templates = []
+      providers.each do |p|
+        templates << p[:templates]
```

Comment:
Same Q as above — p[:templates] seems to be an array, you probably want `+=` to collect a flat array.